Source/Core/Common/Arm64Emitter.h

```diff
@@ -558,156 +534,56 @@ struct LogicalImm
-        // inverted it at the start of the function), which is invalid.
-        return;
-      }
-      else
+      if (value != Common::RotateRight(value, i / 2))
       {
-        // Otherwise, if c was zero but a was not, then there's just one stretch
-        // of set bits in our word, meaning that we have the trivial case of
-        // d == 64 and only one 'repetition'. Set up all the same variables as in
-        // the general case above, and set the N bit in the output.
-        clz_a = Common::CountLeadingZeros(a);
-        d = 64;
-        mask = ~UINT64_C(0);
-        out_n = 1;
+        esize = i;
+        break;
       }
     }
 
-    // If the repeat period d is not a power of two, it can't be encoded.
-    if (!MathUtil::IsPow2<u64>(d))
+    if (value == 0 || (~value) == 0 || esize == 0 || esize > width)
+    {
+      valid = false;
       return;
+    }
 
-    // If the bit stretch (b - a) does not fit within the mask derived from the
-    // repeat period, then fail.
-    if (((b - a) & ~mask) != 0)
-      return;
+    const u64 emask = (~u64{0}) >> (64 - esize);
 
-    // The only possible option is b - a repeated every d bits. Now we're going to
-    // actually construct the valid logical immediate derived from that
-    // specification, and see if it equals our original input.
-    //
-    // To repeat a value every d bits, we multiply it by a number of the form
-    // (1 + 2^d + 2^(2d) + ...), i.e. 0x0001000100010001 or similar. These can
-    // be derived using a table lookup on CLZ(d).
-    constexpr std::array<u64, 6> multipliers = {{
-        0x0000000000000001UL,
-        0x0000000100000001UL,
-        0x0001000100010001UL,
-        0x0101010101010101UL,
-        0x1111111111111111UL,
-        0x5555555555555555UL,
-    }};
-
-    const int multiplier_idx = Common::CountLeadingZeros((u64)d) - 57;
-
-    // Ensure that the index to the multipliers array is within bounds.
-    DEBUG_ASSERT((multiplier_idx >= 0) &&
-                 (static_cast<size_t>(multiplier_idx) < multipliers.size()));
-
-    const u64 multiplier = multipliers[multiplier_idx];
-    const u64 candidate = (b - a) * multiplier;
-
-    // The candidate pattern doesn't match our input value, so fail.
-    if (value != candidate)
-      return;
+    // Extract the repeating element, rotating it such that the LSB is 1:
+    // If LSB is already one, we mask away the trailing sequence of ones and
+    // pick the next sequence of ones. This ensures we get a complete element
+    // that has not been cut-in-half due to rotation across the word boundary.
 
-    // We have a match! This is a valid logical immediate, so now we have to
-    // construct the bits and pieces of the instruction encoding that generates
-    // it.
-    n = out_n;
+    const u64 inverse_mask_from_trailing_ones = ~value | (value + 1);
+    const size_t rotation = Common::LeastSignificantSetBit(value & inverse_mask_from_trailing_ones);
```

Comment:
`LeastSignificantSetBit(x)` can just return `__builtin_ctzll(x)` - the result of which is undefined if x is 0, which it can be in this case. Maybe add a `Common::CountTrailingZeros` matching the zero-handling of `Common::CountLeadingZeros`?

I'd usually clear the trailing ones with just `value & (value + 1)`, but otherwise this looks good.

(In case it's of interest, I wrote a [blog post on optimising logical immediate encoding](https://dougallj.wordpress.com/2021/10/30/bit-twiddling-optimising-aarch64-logical-immediate-encoding-and-decoding/) a while ago, which also removes the initial loop - it'd probably look something like [this](https://gist.github.com/dougallj/6a5594d0040c6f1aa476cb8ca642dfa5).)

Reply:
Ha! Why am I not surprised this isn't a novel insight! Neat!

I intuitively knew you could rotate then mask again, but the observation that verifying rotation also ensured esize was a power of two wasn't obvious to me.

Reply:
Thanks! Yeah, I don't know if that one's obvious to anyone. (I'd consider adding an explicit "is power of two" check to spare future people who stumble upon this code from having to understand that, but it feels silly to add code that I know is redundant.)